tests/internal/cluster/cluster.go

```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+
+	return string(output), err
+}
+
+// ExecCmdWithRetries executes a command on the provided client on each node matching nodeSelector,
+// retrying on internal errors
+// retries times with a 10 second delay between retries, and ignores the stdout.
+func ExecCmdWithRetries(client *clients.Settings, retries uint, nodeSelector, command string) error {
+	for retry := range retries {
+		err := ExecCmd(client, nodeSelector, command)
+		if isErrorExecuting(err) {
+			glog.V(params.LogLevel).Infof("Error during command execution, retry %d (%d max): %w", retry+1, retries, err)
+
+			time.Sleep(10 * time.Second)
+
+			continue
+		}
+
+		return err
+	}
+
+	return fmt.Errorf("ran out of %d retries executing command %s", retries, command)
+}
+
+// ExecCmdWithStdoutWithRetries executes a command on the provided client,
+// retrying on internal errors retries times with a 10
+// second delay between retries, and returns the stdout for each node.
+func ExecCmdWithStdoutWithRetries(
+	client *clients.Settings, retries uint, command string, options ...metav1.ListOptions) (map[string]string, error) {
+	for retry := range retries {
+		outputs, err := ExecCmdWithStdout(client, command, options...)
+		if isErrorExecuting(err) {
+			glog.V(params.LogLevel).Infof("Error during command execution, retry %d (%d max): %w", retry+1, retries, err)
+
+			time.Sleep(10 * time.Second)
+
+			continue
+		}
+
+		return outputs, err
+	}
+
+	return nil, fmt.Errorf("ran out of %d retries executing command %s", retries, command)
+}
+
+// ExecCommandOnSNOWithRetries executes a command on the provided single node client,
+// retrying on internal errors retries times
+// with a 10 second delay between retries, and returns the stdout.
+func ExecCommandOnSNOWithRetries(client *clients.Settings, retries uint, command string) (string, error) {
+	outputs, err := ExecCmdWithStdoutWithRetries(client, retries, command)
+	if err != nil {
+		return "", err
+	}
+
+	if len(outputs) != 1 {
+		return "", fmt.Errorf("expected results from one node, found %d nodes", len(outputs))
+	}
+
+	for _, output := range outputs {
+		return output, nil
+	}
+
+	return "", fmt.Errorf("found unreachable code in ExecCommandOnSNO")
+}
+
+// isErrorExecuting matches errors that contain the message "error executing command in container".
+func isErrorExecuting(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	return strings.Contains(err.Error(), "error executing command in container") ||
+		strings.Contains(err.Error(), "container not found")
+}
+
+// WaitForClusterRecover waits up to timeout for all pods in namespaces on a provided node to recover.
+func WaitForClusterRecover(client *clients.Settings, namespaces []string, timeout, extraWait time.Duration) error {
+	err := waitForClusterReachable(client, timeout)
+	if err != nil {
+		return err
+	}
+
+	err = waitForAllPodsHealthy(client, namespaces, timeout)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(extraWait)
```

Comment:
What is the point of this sleep ?

Reply:
This was to fix instability after reboot when running commands. This is no longer needed with retries. I removed it.